packages/typir/src/utils/utils-definitions.ts

```diff
@@ -29,44 +37,467 @@ export function isNameTypePair(type: unknown): type is NameTypePair {
     return typeof type === 'object' && type !== null && typeof (type as NameTypePair).name === 'string' && isType((type as NameTypePair).type);
 }
 
-// TODO this is a WIP sketch for managing the use of Types in properties/details of other Types (e.g. Types of fields of class Types)
-export interface TypeReference<T extends Type = Type> {
-    readonly ref?: T;
-    readonly selector?: TypeSelector;
-    readonly error?: TypirProblem;
-}
+
 
 // This TypeScript type defines the possible ways to identify a wanted Typir type.
-// TODO find better names
+// TODO find better names: TypeSpecification, TypeDesignation/Designator, ... ?
 export type TypeSelector =
-    | Type      // the instance of the wanted type
-    | string    // identifier of the type (in the type graph/map)
-    | unknown   // domain node to infer the final type from
+    | Type              // the instance of the wanted type
+    | string            // identifier of the type (in the type graph/map)
+    | TypeInitializer   // delayed creation of types
+    | TypeReference     // reference to a (maybe delayed) type
+    | unknown           // domain node to infer the final type from
     ;
-// TODO this is a sketch for delaying the type selection in the future
 export type DelayedTypeSelector = TypeSelector | (() => TypeSelector);
 
-export function resolveTypeSelector(services: TypirServices, selector: TypeSelector): Type {
-    /** TODO this is only a rough sketch:
-     * - detect cycles/deadlocks during the resolving process
-     * - make the resolving strategy exchangable
-     * - integrate it into TypeReference implementation?
+
+export interface WaitingForIdentifiableAndCompletedTypeReferencesListener<T extends Type = Type> {
+    onFulfilled(waiter: WaitingForIdentifiableAndCompletedTypeReferences<T>): void;
+    onInvalidated(waiter: WaitingForIdentifiableAndCompletedTypeReferences<T>): void;
+}
+
+/**
+ * The purpose of this class is to inform its listeners, when all given TypeReferences reached their specified initialization state (or a later state).
+ * After that, the listeners might be informed multiple times,
+ * if at least one of the TypeReferences was unresolved/invalid, but later on all TypeReferences are again in the desired state, and so on.
+ */
+export class WaitingForIdentifiableAndCompletedTypeReferences<T extends Type = Type> implements TypeReferenceListener, TypeStateListener {
```

Comment:
I am happy that your approach is working, but this event emitting back and forth is not so obvious and hard to grasp. Maybe I get how it work when I debug the tests.

I also was thinking about some kind of scheduler like in an operating system. Having structures that are known from university might help to simplify the resolution. But saying is easier than doing.